encoder.py

```diff
@@ -0,0 +1,145 @@
+        ``densenet161``, ``inception_v3``, ``googlenet``,
+        ``shufflenet_v2_x1_0``, ``mobilenet_v2``, ``resnext50_32x4d``,
+        ``wide_resnet50_2``, ``mnasnet1_0``
+    :param pool_strategy: the pooling strategy. Options are:
+        - `None`: Means that the output of the model will be the 4D tensor
+            output of the last convolutional block.
+        - `mean`: Means that global average pooling will be applied to the
+            output of the last convolutional block, and thus the output of
+            the model will be a 2D tensor.
+        - `max`: Means that global max pooling will be applied.
+    :param channel_axis: The axis of the color channel, default is 1
+    :param device: Which device the model runs on. Can be 'cpu' or 'cuda'
+    :param load_pre_trained_from_path: Loads your own model weights form the path. If not provided, the default
+           model will be downloaded from torch hub.
+    :param default_traversal_path: Used in the encode method an define traversal on the received `DocumentArray`
+    :param default_batch_size: Defines the batch size for inference on the loaded PyTorch model.
+    :param args:  Additional positional arguments
+    :param kwargs: Additional keyword arguments
+    """
+    DEFAULT_TRAVERSAL_PATH = 'r'
+
+    def __init__(
+        self,
+        model_name: str = 'mobilenet_v2',
+        pool_strategy: str = 'mean',
+        channel_axis: int = 1,
+        device: Optional[str] = None,
+        load_pre_trained_from_path: Optional[str] = None,
+        default_traversal_path: Optional[str] = None,
+        default_batch_size: Optional[int] = 32,
+        *args,
+        **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+
+        if not device:
+            device = 'cuda' if torch.cuda.is_available() else 'cpu'
+        self.device = device
+        self.default_batch_size = default_batch_size
+
+        self.default_traversal_path = self.DEFAULT_TRAVERSAL_PATH if default_traversal_path is None\
+            else default_traversal_path
+
+        self.channel_axis = channel_axis
+        # axis 0 is the batch
+        self._default_channel_axis = 1
+        self.model_name = model_name
+        if pool_strategy not in ('mean', 'max'):
+            raise NotImplementedError(f'unknown pool_strategy: {self.pool_strategy}')
+        self.pool_strategy = pool_strategy
+
+        if load_pre_trained_from_path:
+            model = getattr(models, self.model_name)(pretrained=False)
+            model.load_state_dict(torch.load(load_pre_trained_from_path))
+        else:
+            model = getattr(models, self.model_name)(pretrained=True)
+
+        self._extract_feature_from_torch_module(model)
+        self.model.to(torch.device(self.device))
+        if self.pool_strategy is not None:
+            self.pool_fn = getattr(np, self.pool_strategy)
+
+    def _extract_feature_from_torch_module(self, model: nn.Module):
```

Comment:
please do it in a functional way

Reply:
Yep, https://github.com/jina-ai/executor-image-torch-encoder/pull/1/commits/1fbaad71e5d35768a2d8db6d0fe2f20633c5b6ae